packages/hardhat/contracts/claim/Staking.sol

```diff
@@ -0,0 +1,234 @@
+    event Unstaked(address indexed user, uint256 amount);
+    event NFTMinted(address indexed user, uint256 numNFTs);
+    event FeesAdjusted(address indexed user, uint256 newFeeLevel);
+
+    constructor(address _softToken, address _softMfersAddress) {
+        softToken = IERC20(_softToken);
+        softMfersContract = ISoftMfers(_softMfersAddress);
+    }
+
+    function getStakedAmount(address user) public view returns (uint256) {
+        return stakedTokens[user].amount;
+    }
+
+    function redeemNFTs() external {
+        uint256 totalSecondsStaked = stakedTokens[msg.sender]
+            .totalStakingDuration;
+
+        if (stakedTokens[msg.sender].amount > 0) {
+            totalSecondsStaked +=
+                block.timestamp -
+                stakedTokens[msg.sender].stakingTime;
+        }
+
+        uint256 totalMonthsStaked = totalSecondsStaked / (30 days);
+        uint256 stakedAmount = stakedTokens[msg.sender].amount;
+        uint256 nftsToRedeem;
+
+        if (stakedAmount >= 1000e18 && stakedAmount < 2500e18) {
+            nftsToRedeem = totalMonthsStaked <= 3
+                ? totalMonthsStaked - stakedTokens[msg.sender].nftsMinted
+                : 3 - stakedTokens[msg.sender].nftsMinted;
+        } else if (stakedAmount >= 2500e18 && stakedAmount < 5000e18) {
+            nftsToRedeem = totalMonthsStaked <= 6
+                ? totalMonthsStaked - stakedTokens[msg.sender].nftsMinted
+                : 6 - stakedTokens[msg.sender].nftsMinted;
+        } else if (stakedAmount >= 5000e18 && stakedAmount < 10000e18) {
+            nftsToRedeem = totalMonthsStaked <= 9
+                ? totalMonthsStaked - stakedTokens[msg.sender].nftsMinted
+                : 9 - stakedTokens[msg.sender].nftsMinted;
+        } else if (stakedAmount >= 10000e18) {
+            nftsToRedeem = totalMonthsStaked <= 12
+                ? totalMonthsStaked - stakedTokens[msg.sender].nftsMinted
+                : 12 - stakedTokens[msg.sender].nftsMinted;
+        }
+
+        require(nftsToRedeem > 0, "No new NFTs to redeem");
+
+        stakedTokens[msg.sender].nftsMinted += nftsToRedeem;
+        softMfersContract.redeem(msg.sender, nftsToRedeem);
+    }
+
+    function stake(
+        uint256 amount,
+        address tokenAddress
+    ) external whenNotPaused {
+        require(amount > 0, "Amount must be greater than 0");
+        require(tokenAddress != address(0), "Invalid token address");
+
+        IERC20 token = IERC20(tokenAddress);
+        uint256 userBalance = token.balanceOf(msg.sender);
+
+        // uint256 userBalance = softToken.balanceOf(msg.sender);
+        require(userBalance >= amount, "Insufficient token balance");
+
+        uint256 userAllowance = token.allowance(msg.sender, address(this));
+        require(userAllowance >= amount, "Insufficient token allowance");
+
+        bool transferSuccess = token.transferFrom(
+            msg.sender,
+            address(this),
```

Comment:
I think we need to alter the design here to have the staked tokens transferred to the governor, not the staking contract.

Reply:
yes good call, I will do that

Reply:
there's definitely certain cons to transferring to the governor but happy to do so